src/language/extension-types.md

```diff
@@ -0,0 +1,315 @@
+extension type E(int i) {
+  E.n(this.i, String foo);
+}
+
+void main() {
+  var named = E.n(3, "Hello");
+  var unnamed = E(4);
+}
+```
+
+### Members
+
+Declare members in the body of an extension type to define its interface. 
+Extension type member declarations are identical to class members.
+Members can be methods, getters, setters, or operators:
+
+```dart
+extension type NumberE(int value) {
+  // Operator:
+  NumberE operator +(NumberE other) =>
+      NumberE(value + other.value);
+  // Getter:
+  NumberE get i => value;
+  // Method:
+  bool isValid() => !value.isNegative;
+}
+```
+
+Members of the representation type are not available to the extension type
+by default. If you want a member from the representation type to be available
+to the extension type, you must write a declaration for it
+in the extension type definition, like the `operator +` in `NumberE`.
+You also can define new members unrelated to the representation type,
+like the `i` getter and `isValid` method.
+
+Extension types can't declare instance fields (unless [`external`][]),
+or [abstract members][].
+
+### Implements
+
+You can optionally use the `implements` clause to introduce
+a subtype relationship on an extension type, making *all* members
+of the representation type available to the extension type:
+
+```dart
+extension type NumberI(int i) 
+  implements int{
+  // ...
+}
+```
+
+An extension type can only implement its representation type
+or a supertype of its representation type.
+
+This is technically NOT an inheritance relationship
+like the one `implements` introduces for classes,
+but rather an [applicability][] relationship like that between an
+extension method and its `on` type.
+Members that apply to the supertype are applicable to the subtype,
+unless the supertype has a more specific definition for the member.
+
+#### `@redeclare`
+
+Replacing a method that shares a name with a member of the supertype
+is not an override relationship for extension types, but a *redeclaration*.
+Use the `@redeclare` annotation to tell the compiler you are *knowingly* choosing
+to use the same name as a supertype's member definition, not just accidentally
+using the same name:
+
+```dart
+class C {
+  void m() {}
+}
+
+extension type E(C c) implements C {
+  @redeclare
+  void m() {}
+}
+```
+
+If you don't use the annotation, the compiler will choose the [more specific][]
+definition wherever the name is invoked, which might not be the behavior you expect.
+
+## Usage
+
+To use an extension type, create an instance the same as you would with a class:
+
+```dart
+extension type NumberE(int value) {
+  NumberE operator +(NumberE other) =>
+      NumberE(value + other.value);
+
+  NumberE get i => value;
+  bool isValid() => !value.isNegative;
+}
+
+void testE() { 
+  var num = NumberE(1);
+}
+```
+
+Then you can invoke members on the object as you would with a class object.
+Any attempt to treat the extension type object as an object
```

Comment:
"Any attempt" is a little ambitious.
I'd usually include `o as RepresentationType` in ways to treat an object as `RepresentationType, and that'll work fine.
Is this trying to say: *Can't assign extension type to representation type, or invoke representation type members on extension type, unless the extension type implements the representation type. Can't assign representation type to extension type ever (unless representation type is `dynamic`).

If that's the point, then I think it has been mostly made above already, but summarizing can be useful.

Maybe:
> Subtyping and assignability works normally for extension
> types, which are (statically) treated as completely new types,
> subtypes of `Object?` and supertypes of `Never`, 
> and otherwise only related to other types through their
> `implements` clauses.
> An extension type cannot be assigned to its representation
> type, and it doesn't expose the representation type's 
> members, not unless the extension type explicitly implements
> the representation type to make itself be a subtype of the 
> representation type (see [transparency](#transparency)). 
>
> Likewise, the representation type of an extension type
> is never a subtype of the extension type, so it cannot
> be used where the extension type is needed.

Reply:
> If that's the point, then I think it has been mostly made above already,

That is the point, and I think you're right. There's a lot of ideas repeated in different sections of this doc I should go through and minimize.